pkg/workceptor/workceptor.go

```diff
@@ -678,25 +688,91 @@ func (w *Workceptor) unitStatusForCFR(unitID string) (map[string]interface{}, er
 	return retMap, nil
 }
 
+func (w *Workceptor) attemptRemoteAction(remoteNodeID, remoteUnitID, remoteAction string, unit *workUnit) error {
+	timeout := time.After(300 * time.Second)
+	retryInterval := 500 * time.Millisecond
+	shouldTimeout := func() bool {
+		select {
+		case <-timeout:
+			logger.Error("%s of remote work unit timed out", remoteAction)
+			return true
+		default:
+			time.Sleep(retryInterval)
+			return false
+		}
+	}
+
+	for {
+		conn, _, err := w.connectToRemote(remoteNodeID)
+		if err != nil {
+			logger.Error("Connection failed to %s: %s\n", remoteNodeID, err)
+		}
+
+		if conn != nil {
+			_, err = conn.Write([]byte(fmt.Sprintf("work %s %s\n", remoteAction, remoteUnitID)))
+			if err != nil {
+				logger.Error("Write error sending to %s: %s\n", remoteNodeID, err)
+			} else {
+				conn.Close()
+				break
+			}
+			conn.Close()
+		}
+
+		if shouldTimeout() {
+			return nil
+		}
+	}
+
+	if remoteAction == "cancel" {
+		var state int
+		for {
+			w.activeUnitsLock.RLock()
+			state = unit.status.State
+			w.activeUnitsLock.RUnlock()
+			if state == WorkStateSucceeded || state == WorkStateFailed {
+				return nil
+			}
+
+			if shouldTimeout() {
```

Comment:
It's a bit weird that we're calling a function named `shouldTimeout` that is actually doing the timeout.  Maybe it should be called `doTimeout`, or just be unrolled into this loop and not be a function at all.  (Also I think most of this logic goes away if we use a Context instead.)